File: src/main/java/com/gcu/userservice/data/UserDataService.java
package com.gcu.userservice.data;

import com.gcu.userservice.data.entity.UserEntity;
import com.gcu.userservice.data.repository.UsersRepository;
import com.gcu.userservice.utility.DatabaseException;
import com.gcu.userservice.utility.NotSupportedException;
import org.springframework.beans.factory.annotation.Autowired;
import java.util.List;
import java.util.Optional;

public class UserDataService implements DataAccessInterface<UserEntity> {

    @Autowired
    private UsersRepository usersRepository;

    public List<UserEntity> findAll(){
        try{
            //Try and get all users from the database
            return usersRepository.findAll();
        } catch (Exception e){
            //Throw a database exception in the event of an error
            throw new DatabaseException();
        }
    }

    @Override
    public List<UserEntity> findAllWithID(String id) {
        throw new NotSupportedException();
    }

    @Override
    public Optional<UserEntity> findByID(String id) {
        try{
            //Try and find user in database with specified ID
            return usersRepository.findById(id);
        } catch (Exception e){
            //Throw a database exception in the event of an error
            throw new DatabaseException();
        }
    }

    @Override
    public UserEntity findBy(UserEntity userEntity) {
        try{
            //Try and find a user by their username and password to authenticate their login credentials
            return usersRepository.getUserEntityByUsernameAndPassword(userEntity.getUsername(), userEntity.getPassword());
        } catch (Exception e){
            //Throw a database exception in the event of an error
            throw new DatabaseException();
        }
    }

    @Override
    public UserEntity findByString(String search) {
        try {
            //Try and find a user by their username or email for the purpose of letting another user start a conversation with them
            return usersRepository.getUserEntityByUsernameOrEmail(search, search);
        } catch (Exception e){
            //Throw a database exception in the event of an error
            throw new DatabaseException();
        }
    }

    @Override
    public boolean create(UserEntity userEntity) {
        try {
            //Try and create a new user entry in the database and then evaluate success by checking if the database assigned the user an ID
            return usersRepository.insert(userEntity).getId().equals("");
        } catch (Exception e){
            //Throw a database exception in the event of an error
            throw new DatabaseException();
        }
    }

    @Override
    public int update(UserEntity userEntity) {
        throw new NotSupportedException();
    }

    @Override
    public int delete(UserEntity userEntity) {
        throw new NotSupportedException();
    }
}

File: src/main/kotlin/com/gcu/userservice/data/UserDataService.kt
package com.gcu.userservice.data

import com.gcu.userservice.data.entity.UserEntity
import com.gcu.userservice.data.repository.UsersRepository
import com.gcu.userservice.utility.DatabaseException
import com.gcu.userservice.utility.ItemNotFoundException
import com.gcu.userservice.utility.NotSupportedException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service("userDataService")
class UserDataService(@Autowired private val userRepository: UsersRepository): DataAccessInterface<UserEntity> {


    override fun findAll(): List<UserEntity> {
        return try {
            userRepository.findAll()
        } catch (e: Exception) {
            throw DatabaseException()
        }
    }

    override fun findAllWithID(id: String): List<UserEntity> {
        throw ItemNotFoundException()
    }

    override fun findByID(id: String): UserEntity? {
        return try {
            userRepository.findById(id).orElseGet(null)
        } catch (e :Exception) {
            throw DatabaseException()
        }
    }

    override fun findBy(t: UserEntity): UserEntity? {
        return try {
            userRepository.getUserEntitiesByUsernameAndPassword(t.username, t.password)
        } catch (e: Exception) {
            throw DatabaseException()
        }
    }

    override fun findByString(search: String): UserEntity? {
        return try {
            userRepository.getUserEntitiesByUsernameOrEmail(search, search)
        } catch (e: Exception) {
            throw DatabaseException()
        }
    }

    override fun create(t: UserEntity): Boolean {
        return try {
            userRepository.insert(t)
            true
        } catch (e: Exception) {
            throw DatabaseException()
        }
    }

    override fun update(t: UserEntity): Boolean {
        throw NotSupportedException()
    }

    override fun delete(t: UserEntity): Boolean {
        throw NotSupportedException()
    }
}